src/quantumchannel.jl

```diff
@@ -0,0 +1,85 @@
+"""
+QuantumChannel for transmitting qubits as `RegRef`s from one register to another in a quantum protocol simulation, 
+with a channel delay and under the influence of a background process.
+The function `put!` is used to put the `RegRef` containing the qubit in the channel, which can then be received by
+the receiving register after a specified delay using the take! method in a synchronous way.
+
+```jldoctest
+julia> using QuantumSavory; using ResumableFunctions; using ConcurrentSim
+
+julia> bell = (Z1⊗Z1 + Z2⊗Z2)/sqrt(2.0); regA = Register(2); regB = Register(2); initialize!((regA[1], regB[2]), bell);
+
+julia> sim = Simulation();
+
+julia> queue = DelayQueue{Register}(sim, 10.0)
+DelayQueue{Register}(QueueStore{Register, Int64}, 10.0)
+
+julia> qc = QuantumChannel(queue)
+QuantumChannel(Qubit(), DelayQueue{Register}(QueueStore{Register, Int64}, 10.0), nothing)
+
+julia> @resumable function alice_node(env, qc)
+            println("Putting Alice's qubit in the channel at ", now(env))
+            put!(qc, regA[1])
+        end
+alice_node (generic function with 1 method)
+
+julia> @resumable function bob_node(env, qc)
+            @yield @process take!(env, qc, regB[1]) # wait for the process with delay to complete
+            println("Taking the qubit from alice at ", now(env))
+        end
+bob_node (generic function with 1 method)
+
+julia> @process alice_node(sim, qc); @process bob_node(sim, qc);
+
+julia> run(sim)
+Putting Alice's qubit in the channel at 0.0
+Taking the qubit from alice at 10.0
+
+julia> regA
+Register with 2 slots: [ Qubit | Qubit ]
+  Slots:
+    nothing
+    nothing
+
+julia> regB
+Register with 2 slots: [ Qubit | Qubit ]
+  Slots:
+    Subsystem 1 of QuantumOpticsBase.Ket 12382959472027850978
+    Subsystem 2 of QuantumOpticsBase.Ket 12382959472027850978
+```
+"""
+struct QuantumChannel{T}
+  trait::T
+  queue::ConcurrentSim.DelayQueue{Register}
+  background::Any
+end
+
+QuantumChannel(queue::ConcurrentSim.DelayQueue{Register}, background=nothing, trait=Qubit()) = QuantumChannel(trait, queue, background)
+
+QuantumChannel(env::ConcurrentSim.Simulation, delay, background=nothing, trait=Qubit()) = QuantumChannel(ConcurrentSim.DelayQueue{Register}(env, delay), background, trait)
+Register(qc::QuantumChannel) = Register([qc.trait], [qc.background])
+
+function Base.put!(qc::QuantumChannel, rref::RegRef)
+  if !isnothing(qc.background)
+      uptotime!(rref, ConcurrentSim.now(qc.queue.store.env))
+  end
+
+  channel_reg = Register(qc)
+  swap!(rref, channel_reg[1])
+
+  if !isnothing(qc.background)
+      uptotime!(channel_reg[1], qc.queue.delay)
+  end
+
+  put!(qc.queue, channel_reg)
+end
+
+
+@resumable function Base.take!(env, qc::QuantumChannel, rref_rec::RegRef)
```

Comment:
Your current `take` is `@resumable` which means you need `@yield @process` instead of just `@yield`. That is not what the API is supposed to be. You already did a few experiments related to this and we have already discussed it in person. After struggling with it for a bit, I think I found the proper way to do it and it is very similar to how `latency` is implemented for the `ConcurrentSim.DelayQueue`.

Basically you need
```
@resumable function post_take(env, take_event)
    item = @yield take_event
    modify_item = -item
    return modify_item
end

function Base.take!(s::YourStoreType)
    inner_take_event = take!(s.q)
    @process post_take(s.q.store.env, inner_take_event)
end
```

Here is a very detailed full example:
```
using ConcurrentSim
using ResumableFunctions

##

struct C
    q::DelayQueue
end

C(env, d) = C(DelayQueue(env, d))

function Base.put!(s, i)
    @info "entering put! $i at $(now(s.q.store.env))"
    put!(s.q, i)
end

@resumable function post_take(env, take_event)
    item = @yield take_event
    modify_item = -item
    @info "post_take got $item at $(now(env)) and modified it into $modify_item"
    return modify_item
end

function Base.take!(s)
    @info "entering take! at $(now(s.q.store.env))"
    inner_take_event = take!(s.q)
    @process post_take(s.q.store.env, inner_take_event)
end

##

sim = Simulation()
c = C(sim, 100.0)

@resumable function producer(sim)
    @yield timeout(sim, 0.1)
    for i in 1:10
        @yield timeout(sim, 1)
        @info "putting $i at $(now(sim))"
        put!(c, i)
        @info "put $i"
    end
end

@resumable function consumer(sim)
    @yield timeout(sim, 20.1)
    for i in 1:10
        @info "taking attempt ($i) at $(now(sim))"
        event = take!(c)
        @info "take event $event made, now yielding"
        content = @yield event
        @info "take yield gives $content at $(now(sim))"
    end
end

@process producer(sim)
@process consumer(sim)

run(sim)
```

Reply:
This is really neat, thanks!